# BEE 4750 Homework 4: Linear Programming and Capacity Expansion

**Name**: Sarah Cooke, Fabien De Silva, Laura Akey

**ID**: sec275, fmd48, lka27

> **Due Date**
>
> Thursday, 11/07/23, 9:00pm

## Overview

### Instructions

-   Problem 1 asks you to formulate and solve a resource allocation
    problem using linear programming.
-   Problem 2 asks you to formulate, solve, and analyze a standard
    generating capacity expansion problem.
-   Problem 3 (5750 only) asks you to add a CO<sub>2</sub> constraint to
    the capacity expansion problem and identify changes in the resulting
    solution.

### Load Environment

The following code loads the environment and makes sure all needed
packages are installed. This should be at the start of most Julia
scripts.

In [5]:
import Pkg
Pkg.activate(@__DIR__)
Pkg.instantiate()

  Activating project at `~/hw04`


## Problems (Total: 50/60 Points)

### Problem 1 (20 points)

A farmer has access to a pesticide which can be used on corn, soybeans,
and wheat fields, and costs \$70/kg-yr to apply. The crop yields the
farmer can obtain following crop yields by applying varying rates of
pesticides to the field are shown in
<a href="#tbl-yields" class="quarto-xref">Table 1</a>.

| Application Rate (kg/ha) | Soybean (kg/ha) | Wheat (kg/ha) | Corn (kg/ha) |
|:------------------------:|:---------------:|:-------------:|:------------:|
|            0             |      2900       |     3500      |     5900     |
|            1             |      3800       |     4100      |     6700     |
|            2             |      4400       |     4200      |     7900     |

Table 1: Crop yields from applying varying pesticide rates for Problem
1.

The costs of production, *excluding pesticides*, for each crop, and
selling prices, are shown in
<a href="#tbl-costs" class="quarto-xref">Table 2</a>.

|   Crop   | Production Cost (\$/ha-yr) | Selling Price (\$/kg) |
|:--------:|:--------------------------:|:---------------------:|
| Soybeans |            350             |         0.36          |
|  Wheat   |            280             |         0.27          |
|   Corn   |            390             |         0.22          |

Table 2: Costs of crop production, excluding pesticides, and selling
prices for each crop.

Recently, environmental authorities have declared that farms cannot have
an *average* application rate on soybeans, wheat, and corn which exceeds
0.8, 0.7, and 0.6 kg/ha, respectively. The farmer has asked you for
advice on how they should plant crops and apply pesticides to maximize
profits over 130 total ha while remaining in regulatory compliance if
demand for each crop (which is the maximum the market would buy) this
year is 250,000 kg?

**In this problem**:

-   Formulate a linear program for this resource allocation problem,
    including clear definitions of decision variable(s) (including
    units), objective function(s), and constraint(s) (make sure to
    explain functions and constraints with any needed derivations and
    explanations). **Tip: Make sure that all of your constraints are
    linear**.
-   Implement the program in `JuMP.jl` and find the solution. How many
    ha should the farmer dedicate to each crop and with what pesticide
    application rate(s)? How much profit will the farmer expect to make?
-   The farmer has an opportunity to buy an extra 10 ha of land. How
    much extra profit would this land be worth to the farmer? Discuss
    why this value makes sense and whether you would recommend the
    farmer should make the purchase.

In [6]:
using JuMP
using HiGHS
using DataFrames
using Plots
using Measures
using CSV
using MarkdownTables

$Problem$ $1$ $Solution$


note: S[0:2], W[0:2], and C[0:2] are in units ha

**Soybeans:**

soybean (kg) = 2900 (kg/ha) *S[0] (ha) + 3800 (kg/ha) *S[1] (ha) + 4400 (kg/ha) *S[2] (ha)

sales_soybeans (dollar) = 0.36 ($/kg) * soybean(kg)

cost_pesticides_soybean (dollar) = 70 ($/kg-yr) * (0*S[0] + 1*S[1] + 2*S[2]) (kg)

cost_production_soybean = 350 * (S[0] + S[1] + S[2]) (kg)


profit_soybean = sales_soybean - cost_pesticides_soybean - cost production soybean

profit_soybean = 0.36(2900*S[0] + 3800*S[1] + 4400*S[2]) - 70(0*S[0] + 1*S[1] + 2*S[2]) - 350(S[0] + S[1] + S[2])

profit_soybean = 1040*S[0] + 1368*S[1] + 1584*S[2] - 70*S[1] - 140*S[2] - 350*S[0] - 350*S[1] - 350*S[2]


**profit_soybean = 690*S[0] + 948*S[1] + 1094*S[2]**

**0.8 * (S[0] + S[1] + S[2]) ≥ (0*S[0] + 1*S[1] + 2*S[2])**


**Wheat:**

wheat (kg) = 3500 (kg/ha) *W[0] (ha) + 4100 (kg/ha) *W[1] (ha) + 4200 (kg/ha) *W[2] (ha)

sales_wheat = 0.27 ($/kg) * wheat(kg)

cost_pesticides_wheat = 70 * (0*W[0] + 1*W[1] + 2*W[2])

cost_production_wheat = 280 * (W[0] + W[1] + W[2])



profit_wheat = sales_wheat - cost_pesticides_wheat - cost production wheat

profit_wheat = 0.27(3500*W[0] + 4100*W[1] + 4200*W[2]) - 70 * (0*W[0] + 1*W[1] + 2*W[2]) -  280 * (W[0] + W[1] + W[2])

profit_wheat = 945*W[0] + 1107*W[1] + 1134*W[2] - 70*W[1] - 140*W[2] - 280*W[0] - 280*W[1] - 280*W[2]


**profit_wheat = 665*W[0] + 757*W[1] + 714* W[2]**

**0.7 * (W[0] + W[1] + W[2]) ≥ (0*W[0] + 1*W[1] + 2*W[2])**


**Corn:**

corn (kg) = 5900 (kg/ha) *C[0] (ha) + 6700 (kg/ha) *C[1] (ha) + 7900 (kg/ha) *C[2] (ha)

profit_corn (dollar) = 0.22($/kg) * corn(kg)

cost_pesticides_corn = 70 * (0*C[0] + 1*C[1] + 2*C[2])

cost_production_corn = 390 * (C[0] + C[1] + C[2])


profit_corn = sales_corn - cost_pesticides_corn - cost_production_corn

profit_corn = 0.22(5900*C[0] + 6700*C[1] + 7900*C[2]) - 70 * (0*C[0] + 1*C[1] + 2*C[2]) - 390 * (C[0] + C[1] + C[2])

profit_corn = 1298*C[0] + 1474*C[1] + 1738*C[2] - 70*C[1] - 140*C[2] - 390*C[0] - 390*C[1] - 390*C[2]


**profit_corn = 908*C[0] + 1014*C[1] + 1208*C[2]**

**0.6 ≥ (0*C[0] + 1*C[1] + 2*C[2]) / (C[0] + C[1] + C[2])**


**Overall:**

130 ≥ S[0] + S[1] + S[2] + W[0] + W[1] + W[2] + C[0] + C[1] + C[2]

250,000 ≥ 2900*S[0] + 3800*S[1] + 4400*S[2] + 3500*W[0] + 4100*W[1] + 4200*W[2] + 5900*C[0] + 6700*C[1] + 7900*C[2]


**Constraints:**

pesticides cost 70 $/kg-yr

average application rate of soybean, wheat, and corn must be below 0.8, 0.7, and 0.6 kg/ha respectively.

soybean ha + wheat ha + corn ha ≤ 130 ha

deamnd for each crop ≤ 250,000 kg/crop

In [27]:
crops_model = Model(HiGHS.Optimizer) # initialize model object
@variable(crops_model, S[0:2] >= 0) # non-negativity constraints
@variable(crops_model, W[0:2] >= 0) # non-negativity constraints
@variable(crops_model, C[0:2] >= 0) # non-negativity constraints
@objective(crops_model, Max, 690*S[0] + 948*S[1] + 1094*S[2] + 665*W[0] + 757*W[1] + 714* W[2] + 908*C[0] + 1014*C[1] + 1208*C[2])
@constraint(crops_model, demand_soy, 2900*S[0]+3800*S[1]+4400*S[2]<=250000)
@constraint(crops_model, demand_wheat, 3500*W[0]+4100*W[1]+4200*W[2]<=250000)
@constraint(crops_model, demand_corn, 5900*C[0]+6700*C[1]+7900*C[2]<=250000)
@constraint(crops_model, soybean, (S[1]+2*S[2]) <= 0.8 * (S[0]+S[1]+S[2]))
@constraint(crops_model, wheat, (W[1]+2*W[2]) <= 0.7 * (W[0]+W[1]+W[2]))
@constraint(crops_model, corn, (C[1]+2*C[2]) <= 0.6 * (C[0]+C[1]+C[2]))
print(crops_model) # this outputs a nicely formatted summary of the model so you can check your specification

Max 690 S[0] + 948 S[1] + 1094 S[2] + 665 W[0] + 757 W[1] + 714 W[2] + 908 C[0] + 1014 C[1] + 1208 C[2]
Subject to
 demand_soy : 2900 S[0] + 3800 S[1] + 4400 S[2] ≤ 250000
 demand_wheat : 3500 W[0] + 4100 W[1] + 4200 W[2] ≤ 250000
 demand_corn : 5900 C[0] + 6700 C[1] + 7900 C[2] ≤ 250000
 soybean : -0.8 S[0] + 0.19999999999999996 S[1] + 1.2 S[2] ≤ 0
 wheat : -0.7 W[0] + 0.30000000000000004 W[1] + 1.3 W[2] ≤ 0
 corn : -0.6 C[0] + 0.4 C[1] + 1.4 C[2] ≤ 0
 S[0] ≥ 0
 S[1] ≥ 0
 S[2] ≥ 0
 W[0] ≥ 0
 W[1] ≥ 0
 W[2] ≥ 0
 C[0] ≥ 0
 C[1] ≥ 0
 C[2] ≥ 0


In [28]:
optimize!(crops_model)

Running HiGHS 1.7.2 (git hash: 5ce7a2753): Copyright (c) 2024 HiGHS under MIT licence terms
Coefficient ranges:
  Matrix [2e-01, 8e+03]
  Cost   [7e+02, 1e+03]
  Bound  [0e+00, 0e+00]
  RHS    [2e+05, 2e+05]
Presolving model
2 rows, 3 cols, 6 nonzeros  0s
1 rows, 2 cols, 2 nonzeros  0s
0 rows, 0 cols, 0 nonzeros  0s
Presolve : Reductions: rows 0(-6); columns 0(-9); elements 0(-18) - Reduced to empty
Solving the original LP from the solution after postsolve
Model   status      : Optimal
Objective value     :  1.4788065362e+05
HiGHS run time      :          0.00


In [29]:
@show value.(S);

value.(S) = 1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, 0:2
And data, a 3-element Vector{Float64}:
 13.812154696132595
 55.248618784530386
  0.0


In [30]:
@show value.(W)

value.(W) = 1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, 0:2
And data, a 3-element Vector{Float64}:
 71.42857142857143
  0.0
  0.0


1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, 0:2
And data, a 3-element Vector{Float64}:
 71.42857142857143
  0.0
  0.0

In [31]:
@show value.(C)

value.(C) = 1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, 0:2
And data, a 3-element Vector{Float64}:
 42.3728813559322
  0.0
  0.0


1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, 0:2
And data, a 3-element Vector{Float64}:
 42.3728813559322
  0.0
  0.0

### Problem 2 (30 points)

For this problem, we will use hourly load (demand) data from 2013 in New
York’s Zone C (which includes Ithaca). The load data is loaded and
plotted below in <a href="#fig-demand" class="quarto-xref">Figure 1</a>.

In [1]:
# load the data, pull Zone C, and reformat the DataFrame
NY_demand = DataFrame(CSV.File("data/2013_hourly_load_NY.csv"))
rename!(NY_demand, :"Time Stamp" => :Date)
demand = NY_demand[:, [:Date, :C]]
rename!(demand, :C => :Demand)
demand[:, :Hour] = 1:nrow(demand)

# plot demand
plot(demand.Hour, demand.Demand, xlabel="Hour of Year", ylabel="Demand (MWh)", label=:false)

Next, we load the generator data, shown in
<a href="#tbl-generators" class="quarto-xref">Table 3</a>. This data
includes fixed costs (\$/MW installed), variable costs (\$/MWh
generated), and CO<sub>2</sub> emissions intensity (tCO<sub>2</sub>/MWh
generated).

In [1]:
gens = DataFrame(CSV.File("data/generators.csv"))

Finally, we load the hourly solar and wind capacity factors, which are
plotted in <a href="#fig-cf" class="quarto-xref">Figure 2</a>. These
tell us the fraction of installed capacity which is expected to be
available in a given hour for generation (typically based on the average
meteorology).

In [1]:
# load capacify factors into a DataFrame
cap_factor = DataFrame(CSV.File("data/wind_solar_capacity_factors.csv"))

# plot January capacity factors
p1 = plot(cap_factor.Wind[1:(24*31)], label="Wind")
plot!(cap_factor.Solar[1:(24*31)], label="Solar")
xaxis!("Hour of the Month")
yaxis!("Capacity Factor")

p2 = plot(cap_factor.Wind[4344:4344+(24*31)], label="Wind")
plot!(cap_factor.Solar[4344:4344+(24*31)], label="Solar")
xaxis!("Hour of the Month")
yaxis!("Capacity Factor")

display(p1)
display(p2)

You have been asked to develop a generating capacity expansion plan for
the utility in Riley County, NY, which currently has no existing
electrical generation infrastructure. The utility can build any of the
following plant types: geothermal, coal, natural gas combined cycle gas
turbine (CCGT), natural gas combustion turbine (CT), solar, and wind.

While coal, CCGT, and CT plants can generate at their full installed
capacity, geothermal plants operate at maximum 85% capacity, and solar
and wind available capacities vary by the hour depend on the expected
meteorology. The utility will also penalize any non-served demand at a
rate of \$10,000/MWh.

**In this problem**:

-   Formulate a linear program for this capacity expansion problem,
    including clear definitions of decision variable(s) (including
    units), objective function(s), and constraint(s) (make sure to
    explain functions and constraints with any needed derivations and
    explanations).
-   Implement your linear program in `JuMP.jl`. Find the optimal
    solution. How much should the utility build of each type of
    generating plant? What will the total cost be? How much energy will
    be non-served?
-   What fraction of annual generation does each plant type produce? How
    does this compare to the breakdown of built capacity that you found
    in Problem 1.5? Do these results make sense given the generator
    data?
-   Make a plot of the electricity price in each hour. Discuss any
    trends that you see.

> **Significant Digits**
>
> Use `round(x; digits=n)` to report values to the appropriate
> precision! If your number is on a different order of magnitude and you
> want to round to a certain number of significant digits, you can use
> `round(x; sigdigits=n)`.

> **Getting Variable Output Values**
>
> `value.(x)` will report the values of a `JuMP` variable `x`, but it
> will return a special container which holds other information about
> `x` that is useful for `JuMP`. This means that you can’t use this
> output directly for further calculations. To just extract the values,
> use `value.(x).data`.

> **Suppressing Model Command Output**
>
> The output of specifying model components (variable or constraints)
> can be quite large for this problem because of the number of time
> periods. If you end a cell with an `@variable` or `@constraint`
> command, I *highly* recommend suppressing output by adding a
> semi-colon after the last command, or you might find that your
> notebook crashes.

### Problem 3 (10 points)

**This problem is only required for students in BEE 5750**.

The NY state legislature is considering enacting an annual
CO<sub>2</sub> limit, which for the utility would limit the emissions in
its footprint to 1.5 MtCO<sub>2</sub>/yr.

**In this problem**:

-   Reformulate your linear program from Problem 2 with any necessary
    changes to capture the CO<sub>2</sub> limit.
-   Implement the new optimization problem and find the optimal
    solution. How much should the utility build of each type of
    generating plant? What is different from your plan from Problem 1?
    Do these changes make sense?
-   What would the value to the utility be of allowing it to emit an
    additional 1000 tCO<sub>2</sub>/yr? An additional 5000?

## References

List any external references consulted, including classmates.